leapsectz/leapsectz.go

```diff
@@ -20,54 +20,109 @@ limitations under the License.
 package leapsectz
 
 import (
+	"bytes"
 	"encoding/binary"
 	"errors"
 	"io"
+	"io/ioutil"
 	"os"
 	"time"
 )
 
 const file = "/usr/share/zoneinfo/right/UTC"
 
 var errBadData = errors.New("malformed time zone information")
+var errBadVersion = errors.New("version is not supported in file")
 
 // LeapSecond represents a leap second
 type LeapSecond struct {
 	Tleap uint32
-	Nleap uint32
+	Nleap int32
+}
+
+// LeapSecond represents a leap second
+type LeapSecondV2 struct {
+	Tleap uint64
+	Nleap int32
+}
+
+// Header represents file header structure. Fields names are copied from doc
+type Header struct {
+	// A four-octet unsigned integer specifying the number of UTC/local indicators contained in the body.
+	isutccnt uint32
+	// A four-octet unsigned integer specifying the number of standard/wall indicators contained in the body.
+	isstdcnt uint32
+	// A four-octet unsigned integer specifying the number of leap second records contained in the body.
+	leapcnt uint32
+	// A four-octet unsigned integer specifying the number of transition times contained in the body.
+	timecnt uint32
+	// A four-octet unsigned integer specifying the number of local time type Records contained in the body - MUST NOT be zero.
+	typecnt uint32
+	// A four-octet unsigned integer specifying the total number of octets used by the set of time zone designations contained in the body.
+	charcnt uint32
 }
 
 // Time returns when the leap second event occurs
 func (l LeapSecond) Time() time.Time {
-	return time.Unix(int64(l.Tleap-l.Nleap+1), 0)
+	if l.Nleap < 0 {
+		return time.Unix(int64(l.Tleap+uint32(l.Nleap)+1), 0)
+	} else {
+		return time.Unix(int64(l.Tleap-uint32(l.Nleap)+1), 0)
+	}
 }
 
 // Parse returns the list of leap seconds
-func Parse() ([]LeapSecond, error) {
-	f, err := os.Open(file)
+func Parse(srcfile string) ([]LeapSecond, []LeapSecondV2, error) {
```

Comment:
Is there any benefit in exposing both lists to the user? I would simplify this API

Reply:
Yeah, I'm using it in Writer

Reply:
Both lists should be in sync and have the same number of leap seconds.

I think we can hide this implementation details to the user and always manage a single list.

In the writer you are modifying both lists and adding the same things, one in 32 bits and one in 64 bits.

In my opinion we should hide all this in the marshalling / unmarshalling of the file and always expose a single list.

In the current implementation we are letting the user set different leap seconds in each list which should not be allowed.

Reply:
Moved to export only one list